src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs

```diff
@@ -34,48 +36,67 @@ public HttpApiHandler(DurableTaskExtension config, TraceWriter traceWriter)
+            this.GetClientResponseLinks(request, instanceId, attribute, out var statusQueryGetUri, out var sendEventPostUri, out var terminatePostUri);
+            return this.CreateCheckStatusResponseMessage(request, instanceId, statusQueryGetUri, sendEventPostUri, terminatePostUri);
+        }
 
-            string taskHub = WebUtility.UrlEncode(attribute.TaskHub ?? config.HubName);
-            string connection = WebUtility.UrlEncode(attribute.ConnectionName ?? config.AzureStorageConnectionStringName ?? ConnectionStringNames.Storage);
 
-            string querySuffix = $"{TaskHubParameter}={taskHub}&{ConnectionParameter}={connection}";
-            if (!string.IsNullOrEmpty(notificationUri.Query))
+        internal async Task<HttpResponseMessage> CreateCheckStatusResponse(
+            HttpRequestMessage request,
+            string instanceId,
+            OrchestrationClientAttribute attribute,
+            TimeSpan timeout,
+            TimeSpan retryInterval)
+        {
+            this.GetClientResponseLinks(request, instanceId, attribute, out var statusQueryGetUri, out var sendEventPostUri, out var terminatePostUri);
+            if (retryInterval > timeout)
             {
-                // This is expected to include the auto-generated system key for this extension.
-                querySuffix += "&" + notificationUri.Query.TrimStart('?');
+                throw new ArgumentException($"Total timeout {timeout.TotalSeconds} should be bigger than retry timeout {retryInterval.TotalSeconds}");
             }
 
-            string statusQueryGetUri = instancePrefix + "?" + querySuffix;
-            string sendEventPostUri = instancePrefix + "/" + RaiseEventOperation + "/{eventName}?" + querySuffix;
-            string terminatePostUri = instancePrefix + "/" + TerminateOperation + "?reason={text}&" + querySuffix;
-
-            HttpResponseMessage response = request.CreateResponse(
-                HttpStatusCode.Accepted,
-                new
+            JToken durableFunctionOutput = null;
+            DurableOrchestrationStatus status;
+            var client = this.GetClient(request);
+            var stopwatch = Stopwatch.StartNew();
+            while (true)
+            {
+                status = await client.GetStatusAsync(instanceId);
+                if (status != null && status.RuntimeStatus == OrchestrationRuntimeStatus.Completed)
                 {
-                    id = instanceId,
-                    statusQueryGetUri = statusQueryGetUri,
-                    sendEventPostUri = sendEventPostUri,
-                    terminatePostUri = terminatePostUri
-                });
-
-            // Implement the async HTTP 202 pattern.
-            response.Headers.Location = new Uri(statusQueryGetUri);
-            response.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(10));
+                    durableFunctionOutput = status.Output;
+                    break;
```

Comment:
I'm having a hard time reasoning about the expected behavior of the code because there are many conditions that I have to keep track of in my head. I think this should be simplified a bit more to reduce complexity. Can we simplify this by removing the `break` statements and replacing them with `return` statements? For example, can we instead `return request.CreateResponse(HttpStatusCode.OK, status.Output);` here?

Reply:
Great point!